Add case-insensitive handling for application security group IDs

snobear opened the pull request:
<!--  All Submissions -->
## Community Note
<!-- Please leave the community note as is. -->
* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review


## Description

This PR adds case-insensitive handling for application security group IDs in both standalone network security rules and inline security rules within network security groups. This resolves issues where Azure Resource Manager IDs with different casing cause unnecessary diffs in Terraform plans.

There is an open PR that is similar (https://github.com/hashicorp/terraform-provider-azurerm/pull/30821), however this PR leverages the existing `DiffSuppressFunc` pattern.

Changes:
- Add DiffSuppressFunc: suppress.CaseDifference to ASG ID fields
- Add custom hash functions for case-insensitive TypeSet hashing

Fixes: Case-only diffs in `azurerm_network_security_rule` and `azurerm_network_security_group` resources when using application security group IDs with inconsistent casing from Azure API responses.

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”


<!-- You can erase any parts of this template below this point that are not applicable to your Pull Request. -->


## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevant documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.


## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

I can’t provide plan/apply logs from our internal terraform unfortunately

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-merging.md).

<!-- Replace the changelog example below with your entry. One resource per line. -->

BUG FIXES:

* `azurerm_network_security_rule` - fix case sensitivity issues with `source_application_security_group_ids` and `destination_application_security_group_ids` fields [GH-30913]
* `azurerm_network_security_group` - fix case sensitivity issues with `source_application_security_group_ids` and `destination_application_security_group_ids` fields in inline `security_rule` blocks [GH-30913]


<!-- What type of PR is this? -->
This is a (please select all that apply):

- [x] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [ ] Enhancement
- [ ] Breaking Change


## Related Issue(s)

Fixes #29841

## AI Assistance Disclosure

<!-- 
IMPORTANT!

If you are using any kind of AI/LLM assistance to contribute to the AzureRM provider, this must be disclosed in the pull request.

If this is the case, please check the box below, and include the extent to which AI was used. (e.g. documentation only, code generation, etc.)

If responses to this pull request are/will be generated using AI, disclose this as well.
-->

- [x] AI Assisted - This contribution was made by, or with the assistance of, AI/LLMs

<!-- extent of AI usage can be described here -->

<!-- heimdall_github_prtemplate:grc-pci_dss-2024-01-05 -->

## Rollback Plan

If a change needs to be reverted, we will publish an updated version of the provider.

## Changes to Security Controls

Are there any changes to security controls (access controls, encryption, logging) in this pull request? If so, explain.

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

snobear committed:
Add case-insensitive handling for application security group IDs

This PR adds case-insensitive handling for application security group IDs
in both standalone network security rules and inline security rules within
network security groups. This resolves issues where Azure Resource Manager
IDs with different casing cause unnecessary diffs in Terraform plans.

Changes:
- Add DiffSuppressFunc: suppress.CaseDifference to ASG ID fields
- Add custom hash functions for case-insensitive TypeSet hashing
- Export hash functions for comprehensive unit testing
- Add comprehensive unit tests for case normalization behavior

Fixes: Case-only diffs in azurerm_network_security_rule and
azurerm_network_security_group resources when using application
security group IDs with inconsistent casing from Azure API responses.

Tested with:
- Unit tests covering hash function determinism and consistency
- Unit tests for DiffSuppressFunc behavior with various case scenarios
- Integration testing with real Azure Resource Manager IDs

magodo reviewed (changes requested):
Hi @snobear, thanks for the PR! Overall it looks good to me 👍 

The only minor point is I feel those unit tests are unnecessary given the function is simple enough.

snobear committed:
Remove unnecessary unit tests

snobear commented:
Hi @magodo, sounds good, easy fix :). Unit tests have been removed

snobear commented:
Excellent, thanks for reviewing @magodo. Anything else needed on my end to get it merged?

magodo commented:
@snobear There will be another one to review this PR, once it gets passed, the PR will be forwarded to Hashicorp for the final review/merge.

ms-henglu commented:
Thanks for the PR! LGTM.

WodansSon reviewed (commented):
@snobear, this LGTM, currently running regression tests. Thanks for the contribution to the provider! 🚀

WodansSon commented:
<img width="1202" height="186" alt="image" src="https://github.com/user-attachments/assets/a5d3d39c-a4af-4a75-ba06-9d31c5678bac" />

<details>
<summary>It looks like the `TestAccNetworkSecurityGroup_applicationSecurityGroup` test is still failing</summary>

          # azurerm_network_security_group.test will be updated in-place
          ~ resource "azurerm_network_security_group" "test" {
                id                  = "/subscriptions/*******/resourceGroups/acctestRG-251029101042222824/providers/Microsoft.Network/networkSecurityGroups/acctestnsg-251029101042222824"
                name                = "acctestnsg-251029101042222824"
              ~ security_rule       = [
                  - {
                      - access                                     = "Allow"
                      - destination_address_prefixes               = []
                      - destination_application_security_group_ids = [
                          - "/subscriptions/*******/resourceGroups/ACCTESTRG-251029101042222824/providers/Microsoft.Network/applicationSecurityGroups/ACCTEST-SECOND251029101042222824",
                        ]
                      - destination_port_ranges                    = [
                          - "443",
                          - "80",
                          - "8080",
                          - "8190",
                        ]
                      - direction                                  = "Inbound"
                      - name                                       = "test123"
                      - priority                                   = 100
                      - protocol                                   = "Tcp"
                      - source_address_prefixes                    = []
                      - source_application_security_group_ids      = [
                          - "/subscriptions/*******/resourceGroups/ACCTESTRG-251029101042222824/providers/Microsoft.Network/applicationSecurityGroups/ACCTEST-FIRST251029101042222824",
                        ]
                      - source_port_ranges                         = [
                          - "10000-40000",
                        ]
                        # (5 unchanged attributes hidden)
                    },
                  + {
                      + access                                     = "Allow"
                      + destination_address_prefixes               = []
                      + destination_application_security_group_ids = [
                          + "/subscriptions/*******/resourceGroups/acctestRG-251029101042222824/providers/Microsoft.Network/applicationSecurityGroups/acctest-second251029101042222824",
                        ]
                      + destination_port_ranges                    = [
                          + "443",
                          + "80",
                          + "8080",
                          + "8190",
                        ]
                      + direction                                  = "Inbound"
                      + name                                       = "test123"
                      + priority                                   = 100
                      + protocol                                   = "Tcp"
                      + source_address_prefixes                    = []
                      + source_application_security_group_ids      = [
                          + "/subscriptions/*******/resourceGroups/acctestRG-251029101042222824/providers/Microsoft.Network/applicationSecurityGroups/acctest-first251029101042222824",
                        ]
                      + source_port_ranges                         = [
                          + "10000-40000",
                        ]
                        # (5 unchanged attributes hidden)
                    },
                ]
                tags                = {}
                # (2 unchanged attributes hidden)
            }
</details>

snobear commented:
Hmm. Perhaps there is another data type coming from the azure api that's not accounted for here. Thanks, I'll take a look.

WodansSon commented:
@snobear 
> Hmm. Perhaps there is another data type coming from the azure api that's not accounted for here. Thanks, I'll take a look.

If you look at the output it appears that there is still a case sensitivity issue with the resource group name segment of the resource ID?

What's coming back from Azure API:
```
/subscriptions/*******/resourceGroups/ACCTESTRG-251029101042222824/providers/Microsoft.Network/applicationSecurityGroups/ACCTEST-FIRST251029101042222824
```

What Terraform is expecting:
```
/subscriptions/*******/resourceGroups/acctestRG-251029101042222824/providers/Microsoft.Network/applicationSecurityGroups/acctest-first251029101042222824
```

snobear commented:
I’ve been working through the failing regression tests here, but no luck yet. I’ve been super tied up at work and haven’t been able to spend much time, so if anyone has some free cycles to dig in go for it. Otherwise, I’ll try to get back to it asap.

magodo committed:
Implement `Hash` function for the `security_rule`

The `Hash` function set for the `source_application_security_group_ids` and `destination_application_security_group_ids` guarantees the hash of these ids are case insensitive. Whilst, when calculating the hash of the whole `security_rule` (precisely, the element of the `security_rule` set), by default it will serialize the values of each embedded fields, regardless the `Hash` set on them (see: terraform-plugin-sdk/v2/helper/schema/serialize.go:SerializeValueForHash). Hence, a specific Hash function is needed for the `security_rule`.

magodo committed:
Merge remote-tracking branch 'upstream/main' into feature/case-insensitive-asg-ids

magodo committed:
ci

magodo committed:
update priority hash setting

snobear commented:
@magodo Go for it! Looks promising, thanks so much.

g3raldtan commented:
Hi! Just wanted to share

I followed snobear response on Creating an Azure Support Case and providing my subscription. 

This was their response:

This is a known, by-design NRP behavior -- not a bug or regression. NRP stores all networking resource keys with UPPERCASE Resource Group Name and Resource Name in its internal Key-Value Store (KVS). When an NSG security rule references an ASG, the API GET response returns the ASG's stored key (uppercase), not the customer-submitted casing. Terraform's AzureRM provider performs case-sensitive comparison of property values (specifically destinationApplicationSecurityGroups and sourceApplicationSecurityGroups within security rules), detects a diff between the submitted lowercase IDs and the returned uppercase IDs, and proposes changes on every terraform plan -- creating perpetual drift.

This same issue was investigated in on of our previous cases ("RG name and ASG names are being uppercased") and resolved as known NRP behavior. The prior investigation confirmed: "All networking resources are stored with uppercase Resource Name and uppercase RG Name."
 
Customer Workarounds:
•	Option 1 (Recommended): Use lifecycle { ignore_changes } in Terraform for the ASG reference properties in NSG security rules to suppress drift detection
•	Option 2: Create resources with UPPERCASE names in Terraform to match NRP's stored format (e.g., STN-ASG-KVMGMT-CHN-PRD instead of stn-asg-kvmgmt-chn-prd)
•	Option 3: Use upper() function in Terraform for resource group and resource names when defining ASG references in security rules
Long-term (No NRP action required):
•	Track Terraform AzureRM provider improvements for case-insensitive property comparison (hashicorp/terraform-provider-azurerm #29841)
•	No NRP-side fix is planned -- this is intentional behavior for NRP's internal key storage


Hope this helps anyone else who encounters this issue.

mstroob commented on internal/services/network/network_security_rule_resource.go:
```diff
@@ -510,3 +517,8 @@ func flattenApplicationSecurityGroupIds(groups *[]securityrules.ApplicationSecur
 }
+
+// HashCaseInsensitiveString provides case-insensitive hashing for TypeSet elements
+func HashCaseInsensitiveString(v interface{}) int {
```
Since this function is replicated twice and could be useful elsewhere so I think it makes sense to remove it from the resource files and move it to internal/tf/pluginsdk/hashstring.go as

```
func HashStringInsensitively(input interface{}) int {
	return HashString(strings.ToLower(input.(string)))
}
```

so it can then be referenced in the schema as pluginsdk.HashStringInsensitively

magodo replied:
Done.

mstroob commented on internal/services/network/network_security_group_resource.go:
```diff
@@ -547,3 +557,56 @@ func validateSecurityRule(sgRule map[string]interface{}) error {
+
+// hashNSR implements a hash function for the `security_rule` property,
+// mainly to normalize the casing for `source_application_security_group_ids` and `destination_application_security_group_ids`.
+func hashNSR(input any) int {
```
```suggestion
func hashNetworkSecurityRule(input any) int {
```

mstroob commented on internal/services/network/network_security_group_resource.go:
```diff
@@ -178,6 +187,7 @@ func resourceNetworkSecurityGroup() *pluginsdk.Resource {
 						},
 					},
 				},
+				Set: hashNSR,
```
```suggestion
				Set: hashNetworkSecurityRule,
```
also move this up, so it is not hidden after all the properties in the security_rule

magodo replied:
Done

mstroob reviewed (commented):
@snobear thanks for the contribution, I have made some suggestions

magodo committed:
Update per review comments

magodo committed:
Move Set function above

magodo commented:
@mstroob All comments resolved.

## Test

```shell
Γ¥» TF_ACC=1 go test -v -run='TestAccNetworkSecurityGroup_|TestAccNetworkSecurityRule' ./internal/services/network
=== RUN   TestAccNetworkSecurityGroup_resourceIdentity
=== PAUSE TestAccNetworkSecurityGroup_resourceIdentity
=== RUN   TestAccNetworkSecurityGroup_list_basic

--- PASS: TestAccNetworkSecurityGroup_list_basic (72.04s)
=== RUN   TestAccNetworkSecurityGroup_basic
=== PAUSE TestAccNetworkSecurityGroup_basic
=== RUN   TestAccNetworkSecurityGroup_requiresImport
=== PAUSE TestAccNetworkSecurityGroup_requiresImport
=== RUN   TestAccNetworkSecurityGroup_singleRule
=== PAUSE TestAccNetworkSecurityGroup_singleRule
=== RUN   TestAccNetworkSecurityGroup_update
=== PAUSE TestAccNetworkSecurityGroup_update
=== RUN   TestAccNetworkSecurityGroup_disappears
=== PAUSE TestAccNetworkSecurityGroup_disappears
=== RUN   TestAccNetworkSecurityGroup_withTags
=== PAUSE TestAccNetworkSecurityGroup_withTags
=== RUN   TestAccNetworkSecurityGroup_addingExtraRules
=== PAUSE TestAccNetworkSecurityGroup_addingExtraRules
=== RUN   TestAccNetworkSecurityGroup_augmented
=== PAUSE TestAccNetworkSecurityGroup_augmented
=== RUN   TestAccNetworkSecurityGroup_applicationSecurityGroup
=== PAUSE TestAccNetworkSecurityGroup_applicationSecurityGroup
=== RUN   TestAccNetworkSecurityRule_resourceIdentity
=== PAUSE TestAccNetworkSecurityRule_resourceIdentity
=== RUN   TestAccNetworkSecurityRule_basic
=== PAUSE TestAccNetworkSecurityRule_basic
=== RUN   TestAccNetworkSecurityRule_requiresImport
=== PAUSE TestAccNetworkSecurityRule_requiresImport
=== RUN   TestAccNetworkSecurityRule_disappears
=== PAUSE TestAccNetworkSecurityRule_disappears
=== RUN   TestAccNetworkSecurityRule_addingRules
=== PAUSE TestAccNetworkSecurityRule_addingRules
=== RUN   TestAccNetworkSecurityRule_augmented
=== PAUSE TestAccNetworkSecurityRule_augmented
=== RUN   TestAccNetworkSecurityRule_applicationSecurityGroups
=== PAUSE TestAccNetworkSecurityRule_applicationSecurityGroups
=== CONT  TestAccNetworkSecurityGroup_resourceIdentity
=== CONT  TestAccNetworkSecurityGroup_applicationSecurityGroup
=== CONT  TestAccNetworkSecurityRule_disappears
=== CONT  TestAccNetworkSecurityRule_basic
=== CONT  TestAccNetworkSecurityRule_requiresImport
=== CONT  TestAccNetworkSecurityGroup_augmented
=== CONT  TestAccNetworkSecurityRule_augmented
=== CONT  TestAccNetworkSecurityRule_applicationSecurityGroups
=== CONT  TestAccNetworkSecurityRule_addingRules
=== CONT  TestAccNetworkSecurityGroup_addingExtraRules
=== CONT  TestAccNetworkSecurityGroup_withTags
=== CONT  TestAccNetworkSecurityGroup_disappears
=== CONT  TestAccNetworkSecurityGroup_requiresImport
=== CONT  TestAccNetworkSecurityRule_resourceIdentity
=== CONT  TestAccNetworkSecurityGroup_update
=== CONT  TestAccNetworkSecurityGroup_singleRule
--- PASS: TestAccNetworkSecurityGroup_resourceIdentity (78.78s)
=== CONT  TestAccNetworkSecurityGroup_basic
--- PASS: TestAccNetworkSecurityGroup_requiresImport (85.40s)
--- PASS: TestAccNetworkSecurityGroup_augmented (88.18s)
--- PASS: TestAccNetworkSecurityGroup_singleRule (93.07s)
--- PASS: TestAccNetworkSecurityRule_augmented (95.19s)
--- PASS: TestAccNetworkSecurityGroup_applicationSecurityGroup (95.71s)
--- PASS: TestAccNetworkSecurityRule_resourceIdentity (95.98s)
--- PASS: TestAccNetworkSecurityGroup_disappears (98.66s)
--- PASS: TestAccNetworkSecurityRule_applicationSecurityGroups (99.47s)
--- PASS: TestAccNetworkSecurityRule_addingRules (101.90s)
--- PASS: TestAccNetworkSecurityGroup_addingExtraRules (107.38s)
--- PASS: TestAccNetworkSecurityGroup_update (110.79s)
--- PASS: TestAccNetworkSecurityRule_requiresImport (116.77s)
--- PASS: TestAccNetworkSecurityRule_disappears (120.29s)
--- PASS: TestAccNetworkSecurityGroup_withTags (126.18s)
--- PASS: TestAccNetworkSecurityRule_basic (126.55s)
--- PASS: TestAccNetworkSecurityGroup_basic (66.23s)
PASS
ok  	github.com/hashicorp/terraform-provider-azurerm/internal/services/network	217.200s
```

mstroob commented on internal/tf/pluginsdk/hashstring.go:
```diff
@@ -20,3 +21,8 @@ func HashString(input interface{}) int {
 }
+
+// HashCaseInsensitiveString provides case-insensitive hashing for TypeSet elements
+func HashCaseInsensitiveString(v interface{}) int {
```
```suggestion
func HashStringInsensitively(v interface{}) int {
```

mstroob commented on internal/services/network/network_security_rule_resource.go:
```diff
@@ -139,8 +140,11 @@ func resourceNetworkSecurityRule() *pluginsdk.Resource {
+					Type:             pluginsdk.TypeString,
+					DiffSuppressFunc: suppress.CaseDifference,
+				},
+				Set: pluginsdk.HashCaseInsensitiveString,
```
```suggestion
				Set: pluginsdk.HashStringInsensitively,
```

mstroob commented on internal/services/network/network_security_rule_resource.go:
```diff
@@ -149,8 +153,11 @@ func resourceNetworkSecurityRule() *pluginsdk.Resource {
+					Type:             pluginsdk.TypeString,
+					DiffSuppressFunc: suppress.CaseDifference,
+				},
+				Set: pluginsdk.HashCaseInsensitiveString,
```
```suggestion
				Set: pluginsdk.HashStringInsensitively,
```

mstroob commented on internal/services/network/network_security_group_resource.go:
```diff
@@ -142,15 +146,21 @@ func resourceNetworkSecurityGroup() *pluginsdk.Resource {
+								Type:             pluginsdk.TypeString,
+								DiffSuppressFunc: suppress.CaseDifference,
+							},
+							Set: pluginsdk.HashCaseInsensitiveString,
```
```suggestion
							Set: pluginsdk.HashStringInsensitively,
```

mstroob commented on internal/services/network/network_security_group_resource.go:
```diff
@@ -142,15 +146,21 @@ func resourceNetworkSecurityGroup() *pluginsdk.Resource {
+								Type:             pluginsdk.TypeString,
+								DiffSuppressFunc: suppress.CaseDifference,
+							},
+							Set: pluginsdk.HashCaseInsensitiveString,
```
```suggestion
							Set: pluginsdk.HashStringInsensitively,
```

mstroob reviewed (commented):
Nit: just want to the change the function name to match the naming convention we seem to have eg: parseAbcInsensitively, once that is done this should be good to merge

magodo committed:
rename function

magodo commented:
@mstroob Function renamed per requested.

mstroob reviewed (approved):
thanks @snobear and @magodo LGTM
